File: recipes_3e/apache/hits/hits.py
#!/usr/bin/python
# hits.py: web page hit-counter example

import sys
import os
import cgi
import cookbook

#@ _GET_HIT_COUNT_
def get_hit_count(conn, page_path):
  cursor = conn.cursor()
  cursor.execute('''
                 INSERT INTO hitcount (path,hits)
                 VALUES(%s,LAST_INSERT_ID(1))
                 ON DUPLICATE KEY UPDATE hits = LAST_INSERT_ID(hits+1)
                 ''', (page_path,))
  count = cursor.lastrowid
  cursor.close()
  conn.commit()
  return count
#@ _GET_HIT_COUNT_

# If script is run from the command line, SCRIPT_NAME won't exist; fake
# it by using script name.

if not os.environ.has_key('SCRIPT_NAME'):
  os.environ['SCRIPT_NAME'] = sys.argv[0]

title = "Hit Count Example"

# Print content type header, blank line that separates
# headers from page body, and first part of page

print("Content-Type: text/html")
print("")
print("<html>")
print("<head><title>%s</title></head>" % title)
print("<body>")

conn = cookbook.connect()

print("<p>Page path: %s</p>" % cgi.escape(os.environ['SCRIPT_NAME'], 1))

# Display a hit count

#@ _DISPLAY_HIT_COUNT_
count = get_hit_count(conn, os.environ['SCRIPT_NAME'])
print("<p>This page has been accessed %d times.</p>" % count)
#@ _DISPLAY_HIT_COUNT_

# Use a logging approach to hit recording.  This enables
# the most recent hits to be displayed.

if os.environ.has_key('REMOTE_HOST'):
  host = os.environ['REMOTE_HOST']
elif os.environ.has_key('REMOTE_ADDR'):
  host = os.environ['REMOTE_ADDR']
else:
  host = 'UNKNOWN'

cursor = conn.cursor()
cursor.execute('''
               INSERT INTO hitlog (path, host) VALUES(%s,%s)
               ''', (os.environ['SCRIPT_NAME'], host))
conn.commit()

# Display the most recent hits for the page

cursor.execute('''
               SELECT DATE_FORMAT(t, '%%Y-%%m-%%d %%T'), host
               FROM hitlog
               WHERE path = %s ORDER BY t DESC LIMIT 10
               ''', (os.environ['SCRIPT_NAME'], ))
print('<p>Most recent hits:</p>')
print('<table border="1">')
print('<tr><th>%s</th><th>%s</th></tr>' % ('Date', 'Host'))
for (date, host) in cursor:
  print('<tr><td>%s</td><td>%s</td></tr>' %
        (cgi.escape(date), cgi.escape(host)))
print('</table>')

cursor.close()
conn.close()

print("</body>")
print("</html>")

File: recipes_3e/apache/hits/hits.rb
#!/usr/bin/ruby
# hits.rb: web page hit-counter example

require "cgi"
require "Cookbook"

#@ _GET_HIT_COUNT_
def get_hit_count(dbh, page_path)
  dbh.do("INSERT INTO hitcount (path,hits) VALUES(?,LAST_INSERT_ID(1))
          ON DUPLICATE KEY UPDATE hits = LAST_INSERT_ID(hits+1)",
         page_path)
  return dbh.func(:insert_id)
end
#@ _GET_HIT_COUNT_

# If script is run from the command line, ENV["SCRIPT_NAME"] won't exist;
# fake it by using script name.

ENV["SCRIPT_NAME"] = $0 if ENV["SCRIPT_NAME"].nil?

title = "Hit Count Example"
page = ""

cgi = CGI.new("html4")

dbh = Cookbook.connect

page << cgi.p { "Page path: " + CGI.escapeHTML(ENV["SCRIPT_NAME"].to_s) }

#@ _DISPLAY_HIT_COUNT_
count = get_hit_count(dbh, ENV["SCRIPT_NAME"])
page << cgi.p { "This page has been accessed #{count} times." }
#@ _DISPLAY_HIT_COUNT_

# Use a logging approach to hit recording.  This enables
# the most recent hits to be displayed.

if ENV["REMOTE_HOST"]
  host = ENV["REMOTE_HOST"]
elsif ENV["REMOTE_ADDR"]
  host = ENV["REMOTE_ADDR"]
else
  host = "UNKNOWN"
end

dbh.do("INSERT INTO hitlog (path, host) VALUES(?,?)", ENV["SCRIPT_NAME"], host)

# Display the most recent hits for the page

page << cgi.p { "Most recent hits:" }
table_rows = cgi.tr {
               cgi.th { "Date" } +
               cgi.th { "Host" }
             }
stmt = "SELECT DATE_FORMAT(t, '%Y-%m-%d %T'), host
        FROM hitlog
        WHERE path = ? ORDER BY t DESC LIMIT 10"
dbh.execute(stmt, ENV["SCRIPT_NAME"]) do |sth|
  sth.fetch do |row|
    row.collect! do |val| CGI.escapeHTML(val.to_s) end
    table_rows << cgi.tr {
                    cgi.td { row[0] } +
                    cgi.td { row[1] }
                  }
  end
end
page << cgi.table("border" => "1") { table_rows }

dbh.disconnect

cgi.out {
  cgi.html {
    cgi.head { cgi.title { title } } +
    cgi.body() { page }
  }
}
